Add generated Admin Router API docs (ngindox)

karlkfi opened the pull request:
## High Level Description

- Add `<adminrouter-src>/api/` directory for [Ngindox](https://github.com/karlkfi/ngindox) YAML and HTML
- Add `make api` to generate above API docs
- Add `packages/adminrouter/extra/src/ci/validate-api-docs.sh` for CI validation that API docs are up to date.
- Add inline comments to Admin Router config for API docs

## Related Issues

  - [DCOS_OSS-696](https://jira.mesosphere.com/browse/DCOS_OSS-696) Generate DC/OS API Doc from AdminRouter nginx.conf

## Checklist for all PR's

  - [X] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: New CI script was added to validate that generated docs match the source configs
  - [X] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [X] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (not required for inline components)
  - [X] Test Results: [public-dcos-adminrouter-api/8](https://jenkins.mesosphere.com/service/jenkins/job/public-dcos-adminrouter-api/8)
  - [ ] Code Coverage (if available): [link to code coverage report]

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dcos/dcos/1460)
<!-- Reviewable:end -->

karlkfi commented:
@vespian: Sharing this early to get buy in on the draft.

I still need to add this to CI somehow and figure out how to get the 1.9.x version of the YAML via HTTP so I can generate new HTML for dcos-website.

The HTML generated here is just for manual verification that it looks good and automatic validation that HTML will generate without error. For dcos-website the command to generate HTML will be slightly different to exclude the inline javascript and css (which has to be modified specifically for the dcos-website environment where it is included into another page).

I also plan to add inline comments to the config files in this PR, to correspond to what I did in the example files: https://github.com/karlkfi/ngindox

karlkfi commented:
Here's the rendered HTML from this PR branch:
https://rawgit.com/mesosphere/dcos/karlkfi/adminrouter-ngindox/packages/adminrouter/extra/src/api/nginx.master.html
https://rawgit.com/mesosphere/dcos/karlkfi/adminrouter-ngindox/packages/adminrouter/extra/src/api/nginx.agent.html

karlkfi commented:
OK to test

karlkfi commented:
OK to test

karlkfi commented:
test this please

karlkfi commented:
retest this please

karlkfi commented:
test this please

karlkfi commented:
test this please

karlkfi commented:
New CI build added:
https://jenkins.mesosphere.com/service/jenkins/view/DCOS/job/public-dcos-adminrouter-api/

Unfortunately, the PR webhook triggering doesn't seem to be working...

Example output: 
https://jenkins.mesosphere.com/service/jenkins/view/DCOS/job/public-dcos-adminrouter-api/7/console

karlkfi commented:
test this please

karlkfi commented:
test this please

karlkfi commented:
test this please

karlkfi commented:
test this please

karlkfi commented:
test this please

karlkfi commented:
FINALLY got Jenkins webhooks working.

Apparently the existing webhooks were using a `github-webhook` url, but the "github pull request builder" I'm used to using requires a `ghprbhook` URL....

So I added a new webhook for `ghprbhook` with just PR & Issue Comment events.

karlkfi commented:
rebased

vespian commented:
> I still need to add this to CI somehow and figure out how to get the 1.9.x version of the YAML via HTTP so I can generate new HTML for dcos-website.

Please add it to the existing CI as just an another buildstep/stage - just like i.e. `flake8` check are invoked.

> New CI build added: https://jenkins.mesosphere.com/service/jenkins/view/DCOS/job/public-dcos-adminrouter-api/

Let's please not multiply CIs, AR already has one and there were some discussions already with #foundations about integrating it into DC/OS tests to have just one run for all the stuff.

On top of that the check `jenkins/public-dcos-adminrouter-api` is now failing and confusing people as the code it requires is still not in the master branch.

> Apparently the existing webhooks were using a github-webhook url, but the "github pull request builder" I'm used to using requires a ghprbhook URL....

Because we were not using it - Jenkins Pipeline scripts do not support GitHubPullRequestBuilder hence the plain GH hook address.

We can jump on quick hangout and I can explain to you how to integrate it with existing CI :)
 

---

Reviewed 2 of 7 files at r1, 2 of 7 files at r3.
Review status: 4 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 5 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-wtZwekJ37A0scrmS:-Ki-wtZwekJ37A0scrmT:b-tuexku) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L5)):*
> ```
> 
> # Absolute path to AdminRouter source, even if make is run from elsewhere
> SRC_DIR:=$(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
> ```

Do we use it here somewhere?

---

*[packages/adminrouter/extra/src/Makefile.common, line 98 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-ww3dHFX1CdVhKGnA:-Ki-ww3dHFX1CdVhKGnB:b-q1am7h) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L98)):*
> ```
> .PHONY: api
> api:
> 	@for prefix in nginx.master nginx.agent ; do \
> ```

Soonish we will split these files into multiple smaller ones - this will not be a problem, right?

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki-xtf6dpNqpc-lHaU3:b-2wl03p) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
> ```
> 	@for prefix in nginx.master nginx.agent ; do \
> 		echo "Generating api/$${prefix}.yaml" && \
> 		docker run --rm -v "$(SRC_DIR):/src" karlkfi/ngindox parse -f /src/$${prefix}.conf > $(SRC_DIR)/api/$${prefix}.yaml && \
> ```

Can we include this tool in the devkit container so that we do not have two different development environments for AR?

---

*[packages/adminrouter/extra/src/README.md, line 16 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-yyHKM7rQtSLeo7o9:-Ki-yyHKM7rQtSLeo7oA:bbi9a28) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/README.md#L16)):*
> ```Markdown
> 
> - NGINX: [nginx.master.conf](nginx.master.conf)
> - YAML: [api/nginx.master.yaml](api/nginx.master.yaml)
> ```

The `api` directory may suggest that this directory somehow contains API-related code, while it's just the documentation. Can we move it into `doc/` directory? At some point, I will most probably have to split README.md so this will be a good place to put individual README.mD sections too.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki-zhp9mEf2P2hNyEvM:bs8mxsm) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
> ```Shell
> #!/usr/bin/env bash
> ```

Let's please include it in the `Makefile.in` for now - I do not want to multiply directories we keep around and this script is mall enough, for now, to not to justify individual file IMO. I am thinking about something like `make verify-api-docs`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460#-:-Ki-uCtFdLaEPMZpCGDu:bstckh1)*
<!-- Sent from Reviewable.io -->

karlkfi commented:
Looks like Reviewable doesn't use line comments or the github review system. So there's no way to respond in thread to each comment without using Reviewable.

karlkfi commented:
Reviewable also seems really confused... It says the Makefile.common changes aren't in the PR.
<img width="535" alt="screen shot 2017-04-18 at 9 08 21 am" src="https://cloud.githubusercontent.com/assets/3980984/25140907/b091054a-2416-11e7-8369-73b1b49669f8.png">

vespian commented:
kk, let's drop reviewable then - GH PR discussions can get messy, but reviewable seems to also be problematic :/ I will resubmit the PR comments in GH way.

FYI @adamtheturtle

karlkfi commented:
Review status: 4 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 5 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-wtZwekJ37A0scrmS:-Ki0rw8FaLe89C7AcaOZ:bxwv7s4) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L5)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Do we use it here somewhere?
</blockquote></details>

This is easier to see in Github's interface. It's used in `make api`. 
The primary reason to use this variable is because PWD is the directory of the user, not the directory of the Makefile. This allows executing the makefile from outside the source directly: `make api -f packages/adminrouter/extra/src/Makefile`.

---

*[packages/adminrouter/extra/src/Makefile.common, line 98 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-ww3dHFX1CdVhKGnA:-Ki0sb0gqAprd1_fozJ4:b-e7zrbp) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L98)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Soonish we will split these files into multiple smaller ones - this will not be a problem, right?
</blockquote></details>

Depends what you mean by "problem". It you're planning to use more includes, that's fine. If you're planning to have multiple entrypoints, they'd have to be added here and generate more yaml and html files and we'd need to re-think the docs pages. 
What are you planning?

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki0supvOD_m4nqzcEwF:boxf4gb) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Can we include this tool in the devkit container so that we do not have two different development environments for AR?
</blockquote></details>

Ngindox is not a development environment, it's effectively just an executable in a container image. As of now tho it requires node & npm to install and run. I doubt you want or need either of those in your development environment. That said, it depends on how you use your development environment... Personally I never actually develop in a container, because I prefer a GUI IDE. 

There's also a limitation that Ngindox doesn't actually produce a self-contained executable yet. It executes a `cli.js` as the entrypoint, using node directly. I'd like to fix that, but it's not done yet.

---

*[packages/adminrouter/extra/src/README.md, line 16 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-yyHKM7rQtSLeo7o9:-Ki0p2VHJalDkC0DD_xU:b-y5k2r1) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/README.md#L16)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

The `api` directory may suggest that this directory somehow contains API-related code, while it's just the documentation. Can we move it into `doc/` directory? At some point, I will most probably have to split README.md so this will be a good place to put individual README.mD sections too.
</blockquote></details>

It depends on whether you think the YAML is the API or just documents the API...
I put it in `api/` so that it matched `make api`. They should probably match, whatever we change it to. And I didn't think `make doc` or `make docs` made sense, because there might be more docs than just the api docs.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki0q6f6o27HJ8agYrNJ:b-b2cvj9) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Let's please include it in the `Makefile.in` for now - I do not want to multiply directories we keep around and this script is mall enough, for now, to not to justify individual file IMO. I am thinking about something like `make verify-api-docs`
</blockquote></details>

I hate giant makefiles, they're nearly impossible to maintain by anyone but the author, and rife with "tricks" to get anything complex done, but if that's how you want to do it, fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460)*
<!-- Sent from Reviewable.io -->

vespian commented:
Reviewed 1 of 7 files at r1, 1 of 1 files at r2, 5 of 7 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 97 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki0wAYKv7C4jNlvI3AP:-Ki0wAYKv7C4jNlvI3AQ:b-e9ipv6) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L97)):*
> ```
> 
> .PHONY: api
> api:
> ```

Can we please make it `api-docs`?  Just an`api` can have many different meanings.

---

*[packages/adminrouter/extra/src/Makefile.common, line 98 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-ww3dHFX1CdVhKGnA:-Ki0vNCaOxDuqwc7NGPc:b-dl8fsi) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L98)):*
<details><summary><i>Previously, karlkfi (Karl Isenberg) wrote…</i></summary><blockquote>

Depends what you mean by "problem". It you're planning to use more includes, that's fine. If you're planning to have multiple entrypoints, they'd have to be added here and generate more yaml and html files and we'd need to re-think the docs pages. 
What are you planning?
</blockquote></details>

More includes, entrypoints stay the same. OK, so this is not an issue then.

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki0vuIoeknCIEwtBYHa:b-em4w1h) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
>  I doubt you want or need either of those in your development environment.
I would :)

The `make api-docs` target will become an integral part of the build process. This means that nginxdox should be shared with the rest of the dev environment.

Also, the `devkit` container is meant to provide a development environment where people can immediately start developing AR, this includes the tools you use to generate doc.

It does not make much difference for dcos build process, as the container is cached anyway.

And last but not least - I would very much prefer to have the definition of all the tooling needed for NGINX doc generation in the AR repo itself. ATM I do not know how exactly the `karlfi/ngindoc` tool is built.

---

*[packages/adminrouter/extra/src/README.md, line 16 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-yyHKM7rQtSLeo7o9:-Ki0xPNTLfEIVID3OP5S:b-b124ax) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/README.md#L16)):*
> I put it in api/ so that it matched make api.

We currently have `make test` yet we have the `test-harness/tests/` directory where the tests reside. Should I also rename it? I think that `make api-docs` target plus putting the results in `docs/api/` seems like a good compromise.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki10Qjm9RvYCfa0jNcP:bq74iaz) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
<details><summary><i>Previously, karlkfi (Karl Isenberg) wrote…</i></summary><blockquote>

I hate giant makefiles, they're nearly impossible to maintain by anyone but the author, and rife with "tricks" to get anything complex done, but if that's how you want to do it, fine.
</blockquote></details>

I am really trying to keep things neat and simple. As soon as the Makefile starts being unmanagable, I will break it down into smaller parts like like for dcos-ca (https://github.com/mesosphere/dcos-ca/tree/master/helper_scripts). But let's cross this bridge when we get there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460)*
<!-- Sent from Reviewable.io -->

orsenthil commented:
+1 to dropping reviewable. It does not play well with Github's review. It's situation in which we either use github or use reviewable. It seems using github is less cumbersome now.

karlkfi commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki19AsEZh2TW5kgVF7B:beu0gcj) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

>  I doubt you want or need either of those in your development environment.
I would :)

The `make api-docs` target will become an integral part of the build process. This means that nginxdox should be shared with the rest of the dev environment.

Also, the `devkit` container is meant to provide a development environment where people can immediately start developing AR, this includes the tools you use to generate doc.

It does not make much difference for dcos build process, as the container is cached anyway.

And last but not least - I would very much prefer to have the definition of all the tooling needed for NGINX doc generation in the AR repo itself. ATM I do not know how exactly the `karlfi/ngindoc` tool is built.
</blockquote></details>

karlkfi/ngindox is a standalone tool: https://github.com/karlkfi/ngindox

It is not specific to Admin Router, and can be used to generate docs for any NGINX config. 
I'm not going to put the ngindox source in the dcos repo. That doesn't make sense for a standalone tool.

You also don't need to know how it's built when it's distributed in a container, that's kinda the point. That said, the how is in the readme, if you're interested.

If you want me to install node & npm & ngindox in your devkit, I can do that later. But I can't do it NOW, because ngindox doesn't install as a binary yet. Once I figure that out, using it without Docker will be easier. I'm still not sure I want to be distributing binaries for a node project, or if that's even feasible without requiring node to be installed, which would make it pointless.

Another option is to use docker-in-docker in your devkit.

---

*[packages/adminrouter/extra/src/README.md, line 16 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-yyHKM7rQtSLeo7o9:-Ki1925UkGLZ7-JMeNiS:bpborz) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/README.md#L16)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

> I put it in api/ so that it matched make api.

We currently have `make test` yet we have the `test-harness/tests/` directory where the tests reside. Should I also rename it? I think that `make api-docs` target plus putting the results in `docs/api/` seems like a good compromise.
</blockquote></details>

`make api-docs` and `docs/api/` sounds fine.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki1BREQ85AyyYidUaZ3:b-gawd5e) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

I am really trying to keep things neat and simple. As soon as the Makefile starts being unmanagable, I will break it down into smaller parts like like for dcos-ca (https://github.com/mesosphere/dcos-ca/tree/master/helper_scripts). But let's cross this bridge when we get there.
</blockquote></details>

That's basically exactly what i did, just with a different directory name...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460)*
<!-- Sent from Reviewable.io -->

vespian commented:
KK, Let's just finish this review in Reviewable so that we do not mess things up too much.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki1GTUuKt3rY62rau_r:b-xv27qj) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
> I'm not going to put the ngindox source in the dcos repo. That doesn't make sense for a standalone tool.

Totally, I am *very* far from asking you to do it :)

> You also don't need to know how it's built when it's distributed in a container, that's kinda the point.

All I want is you to add steps from https://github.com/karlkfi/ngindox/blob/master/README.md -> add them to the Dockerfile of the devkit container so that the tool is installed *in* the devkit container. It does not have to be binary - I am also compiling AR in the very same docker container.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki1Fg3x4JPrPm_HetMk:b-gyf661) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
<details><summary><i>Previously, karlkfi (Karl Isenberg) wrote…</i></summary><blockquote>

That's basically exactly what i did, just with a different directory name...
</blockquote></details>

Yep, but as I said - let's cross this bridge when we get there. For now I would relly like to take advantage of the fact that AR is in DC/OS repo and I do not have to resort to `helper_scripts` approach/keep everything in one place for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460#-:-Ki1FMDtek76MeTFLNPo:bwoo430)*
<!-- Sent from Reviewable.io -->

karlkfi commented:
- Moved `api` to `docs/api`
- Renamed `make api` to `make api-docs`
- Replaced `ci/validate-api-docs.sh` with `make check-api-docs`
- Updated `Jenkinsfile-insecure.groovy` to run `make check-api-docs`

karlkfi commented:
I can still put ngindox into the devkit, but I don't think it's actually worth the effort... If I install it in the devkit it makes the usage actually more complicated, because it wouldn't be in the entrypoint...

karlkfi commented:
Alright, I've published ngindox to npm and fixed the install issue. There's no self-contained binary, because that's apparently non-trivial with npm, but it is at least easily installable:

```
$ curl -sL https://deb.nodesource.com/setup_5.x | sudo -E bash -
$ sudo apt-get install -y nodejs
$ npm install -g ngindox
```

karlkfi commented:
Review status: 3 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[packages/adminrouter/extra/src/Makefile.common, line 97 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki0wAYKv7C4jNlvI3AP:-Ki32UAnekftmc5KT-_j:b-896fix) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L97)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Can we please make it `api-docs`?  Just an`api` can have many different meanings.
</blockquote></details>

Done.

---

*[packages/adminrouter/extra/src/Makefile.common, line 100 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-xtf6dpNqpc-lHaU2:-Ki32_AaqIyG0EoDFVSt:b-896fix) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/Makefile.common#L100)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

> I'm not going to put the ngindox source in the dcos repo. That doesn't make sense for a standalone tool.

Totally, I am *very* far from asking you to do it :)

> You also don't need to know how it's built when it's distributed in a container, that's kinda the point.

All I want is you to add steps from https://github.com/karlkfi/ngindox/blob/master/README.md -> add them to the Dockerfile of the devkit container so that the tool is installed *in* the devkit container. It does not have to be binary - I am also compiling AR in the very same docker container.
</blockquote></details>

Done.

---

*[packages/adminrouter/extra/src/ci/validate-api-docs.sh, line 1 at r3](https://reviewable.io:443/reviews/dcos/dcos/1460#-Ki-zhp9mEf2P2hNyEvL:-Ki32aZn6qJsJKqCV7JI:b-896fix) ([raw file](https://github.com/dcos/dcos/blob/2d0687cf561bca5ddbd3d230648ff94d9d90d941/packages/adminrouter/extra/src/ci/validate-api-docs.sh#L1)):*
<details><summary><i>Previously, vespian (Paweł Rozlach) wrote…</i></summary><blockquote>

Yep, but as I said - let's cross this bridge when we get there. For now I would relly like to take advantage of the fact that AR is in DC/OS repo and I do not have to resort to `helper_scripts` approach/keep everything in one place for now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1460)*
<!-- Sent from Reviewable.io -->

karlkfi commented:
- Added Node, NPM, and Ngindox to adminrouter-devkit
- Updated `make api-docs` to use ngindox in adminrouter-devkit

karlkfi commented:
Alright, @vespian. I believe I addressed all your feedback. LMK.

vespian commented:
@adamtheturtle Please have a look - let's ship it soonish!

vespian commented on Jenkinsfile-insecure.groovy:
```diff
@@ -4,6 +4,10 @@ dir("packages/adminrouter/extra/src") {
     }
 
     try {
+        stage('make check-api-docs') {
```
Btw, we will also have to put it into EE: https://github.com/mesosphere/dcos-enterprise/blob/master/Jenkinsfile#L55-L64

karlkfi replied:
If that's the only thing that needs porting, we're in great shape!

karlkfi replied:
I think the EE config files will need updating too tho, to add comments.

vespian replied:
Yeah, we probably should ;) 

The deduplication of nginx.(master|agent).conf is going to happen soonish (<1 month) hopefully, So far almost all lua code got deduplicated.

jgehrcke commented on packages/adminrouter/extra/src/docs/api/nginx.agent.html:
```diff
@@ -0,0 +1,683 @@
+      margin: 0 5px;
+    }
+  </style>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.0/jquery.min.js" integrity="sha256-JAW99MJVpJBGcbzEuXk4Az05s/XyDdBomFqNlM3ic+I=" crossorigin="anonymous"></script>
```
Sweet, I like to see you using the new `integrity` feature.

karlkfi replied:
That was actually in the standard examples on the jquery site! It's pretty neat.

jgehrcke commented on packages/adminrouter/extra/src/docs/api/nginx.master.yaml:
```diff
@@ -0,0 +1,332 @@
+  /:
+    group: Root
+    matcher: path
+    description: DC/OS GUI
```
Is "DC/OS GUI" our latest and greatest official name for the web UI?

karlkfi replied:
That's the name of the component, per https://dcos.io/docs/1.9/overview/architecture/components/

jgehrcke commented on packages/adminrouter/extra/src/docs/api/nginx.master.yaml:
```diff
@@ -0,0 +1,332 @@
+  /acs/api/v1:
+    group: Authentication
+    matcher: path
+    description: Access Control Service
```
'Access Control Service' is a legacy term from Enterprise DC/OS. I think here it might be more correct to say "DC/OS Authentication (OAuth)" (same as above).

karlkfi replied:
Yeah, I'm torn on this one. I wanted to describe the acronym ACS, because it's in the route path. And the description is of the route, not of the backend component...

jgehrcke commented on packages/adminrouter/extra/src/docs/api/nginx.master.yaml:
```diff
@@ -0,0 +1,332 @@
+  /acs/api/v1/auth/:
+    group: Authentication
+    matcher: path
+    description: Access Control Service (unauthenticated)
```
Same here.

jgehrcke commented on packages/adminrouter/extra/src/docs/api/nginx.master.yaml:
```diff
@@ -0,0 +1,332 @@
+  /dcos-metadata/dcos-version.json:
+    group: Metadata
+    matcher: path
+    description: DC/OS version (unauthenticated)
```
Can readers of this documentation to rely on all endpoints that do not say "(unauthenticated)" being authenticated? :)

karlkfi replied:
This is still up for debate... 
I haven't specifically stated that everything is authenticated unless specified, but it seems silly to mark everything else as authenticated. Because authentication is enforced using a custom Lua script, there's no way to know from a generic NGINX perspective what that Lua script does. So it doesn't make sense for Ngindox (which is not dc/os specific) to know that...

So I'm not really sure how to programmatically determine auth-required status per route.

That's why I just specified it via author-defined description string.

karlkfi replied:
I could add a new Ngindox metadata field for `Authentication: required`, but then how should I display it in the UI/HTML? And would it be useful in a non-dcos context?

jgehrcke commented on packages/adminrouter/extra/src/nginx.master.conf:
```diff
@@ -74,11 +97,15 @@ http {
         }
 
+        # Group: Authentication
+        # Description: Access Control Service (unauthenticated)
```
see above

jgehrcke commented on packages/adminrouter/extra/src/nginx.master.conf:
```diff
@@ -74,11 +97,15 @@ http {
         }
 
+        # Group: Authentication
+        # Description: Access Control Service
```
see above

karlkfi commented:
@jgehrcke: Do you have a better suggestion for the descriptions of the following routes?
- `/acs/api/v1`
- `/acs/api/v1/auth/`
Or should we just punt on it?
If not. Are we good to go?

FWIW, the conf file comments are the source of truth. Ngindox uses them to generate the yaml file.

karlkfi commented:
@mesosphere-mergebot bump-ee

karlkfi commented:
Looks like the `Cache` and `Deprecated` metadata fields are getting lost by Ngindox.... need to fix that before merging.

karlkfi commented:
Alright, I fixed the `Cache` and `Deprecated` metadata fields. I think we're good to go!
